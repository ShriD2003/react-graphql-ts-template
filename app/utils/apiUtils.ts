import { ApisauceInstance, create } from 'apisauce';
import snakeCase from 'lodash/snakeCase';
import camelCase from 'lodash/camelCase';
import { mapKeysDeep } from './index';

const apiClients: Record<string, ApisauceInstance> = {};

export const getApiClient = (type = 'github') => apiClients[type];

export const generateApiClient = (type = 'github') => {
  switch (type) {
    case 'github':
      apiClients[type] = createApiClientWithTransForm(process.env.GITHUB_URL!);
      return apiClients[type];
    default:
      apiClients.default = createApiClientWithTransForm(process.env.GITHUB_URL!);
      return apiClients.default;
  }
};

export const createApiClientWithTransForm = (baseURL: string) => {
  const api = create({
    baseURL,
    headers: { 'Content-Type': 'application/json' }
  });
  api.addResponseTransform((response) => {
    const { ok, data } = response;
    if (ok && data) {
      response.data = mapKeysDeep(data, (keys: string) => camelCase(keys));
    }
    return response;
  });

  api.addRequestTransform((request) => {
    const { data } = request;
    if (data) {
      request.data = mapKeysDeep(data, (keys: string) => snakeCase(keys));
    }
    return request;
  });
  return api;
};
