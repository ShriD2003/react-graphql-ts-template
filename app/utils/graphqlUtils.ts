import { Launch } from '@app/containers/HomeContainer';
import ApolloClient, { DocumentNode, InMemoryCache, QueryOptions } from 'apollo-boost';

export const client = new ApolloClient({
  uri: 'https://api.spacex.land/graphql',
  cache: new InMemoryCache()
});

interface ResponseData {
  data: Launch | null;
  ok: boolean | null;
  error: Object | null;
}

export const getQueryResponse = (query: DocumentNode, variables?: QueryOptions['variables']) => {
  const responseData: ResponseData = {
    data: null,
    error: null,
    ok: false
  };
  return client
    .query<ResponseData>({ query: query, variables })
    .then((res) => {
      if (res.errors) {
        return { ...responseData, error: res.errors };
      } else {
        return { ...responseData, data: res, ok: true };
      }
    })
    .catch((err) => {
      return { ...responseData, error: err };
    });
};
