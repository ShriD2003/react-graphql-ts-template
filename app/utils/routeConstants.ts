export type RouteConstant = {
  route: string;
  exact?: boolean;
  isProtected?: boolean;
  props?: object;
};

const routeConstants: Record<string, RouteConstant> = {
  home: {
    route: '/',
    exact: true
  },
  launch: {
    route: '/:launchId',
    exact: true
  }
};

export type RouteKeys = keyof typeof routeConstants;

export default routeConstants;
