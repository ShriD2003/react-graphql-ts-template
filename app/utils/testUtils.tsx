import React from 'react';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Route, Router } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import history from './history';

export const renderWithIntl = (children: React.ReactNode) => {
  i18n.load(DEFAULT_LOCALE, translationMessages[DEFAULT_LOCALE]);
  i18n.activate(DEFAULT_LOCALE);

  return render(<I18nProvider i18n={i18n}> {children} </I18nProvider>);
};

export const getComponentStyles = (Component: React.FC<any>, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props)!.type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  return window.getComputedStyle(componentRoots[0]);
};

export const renderProvider = (children: React.ReactNode, { path }: { path?: string } = {}, renderFn = render) => {
  const store = configureStore({}).store;
  return renderFn(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          <Router history={history}>{path ? <Route path={path}>{children}</Route> : children}</Router>
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};
export const timeout = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = <Data,>(ok: boolean, data: Data, error?: object) => ({
  ok,
  data,
  error
});
