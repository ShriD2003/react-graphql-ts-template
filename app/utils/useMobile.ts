import { useState, useEffect } from 'react';

const useMobile = (breakpoint: number) => {
  const [mobile, setMobile] = useState(window.innerWidth < breakpoint);

  function detectMobile() {
    if (window.innerWidth < breakpoint) {
      setMobile(true);
    } else {
      setMobile(false);
    }
  }

  useEffect(() => {
    window.addEventListener('resize', detectMobile);
    return () => {
      window.removeEventListener('resize', detectMobile);
    };
  }, []);

  return {
    mobile
  };
};

export default useMobile;
