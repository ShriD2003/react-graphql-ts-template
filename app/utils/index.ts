import find from 'lodash-es/find';
import get from 'lodash-es/get';
import history from './history';
import routeConstants from './routeConstants';

export const getCurrentRouteDetails = (location: Partial<Location>) => {
  if (!get(location, 'pathname')) {
    return null;
  }
  const route = find(
    Object.keys(routeConstants),
    (key) => routeConstants[key].route === location.pathname || `${routeConstants[key].route}/` === location.pathname
  );
  if (route) {
    return routeConstants[route];
  }
  return null;
};
export const mapKeysDeep = <T>(obj: T, fn: (key: string) => string): T =>
  Array.isArray(obj)
    ? obj.map((val) => mapKeysDeep(val, fn))
    : typeof obj === 'object'
    ? Object.keys(obj).reduce((acc, current) => {
        const key = fn(current);
        const val = obj[current as keyof T];
        acc[key] = val !== null && typeof val === 'object' ? mapKeysDeep(val, fn) : val;
        return acc;
      }, {} as any)
    : obj;

export const isLocal = () => {
  try {
    if (process.env.IS_LOCAL) {
      const local = JSON.parse(process.env.IS_LOCAL);
      return typeof local === 'boolean' && local;
    }
  } catch {
    // continue regardless of error
  }
  return false;
};

export const setQueryParam = (param: string, value: string | number, op: 'set' | 'delete' = 'set') => {
  const urlParams = new URLSearchParams(history.location.search);
  urlParams[op](param, value as string);
  history.push({ search: urlParams.toString() });
};
