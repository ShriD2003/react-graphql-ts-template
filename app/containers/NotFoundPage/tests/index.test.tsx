import React from 'react';
import { fireEvent } from '@testing-library/react';
import { I18nProvider } from '@lingui/react';
import { i18n } from '@lingui/core';

import NotFoundPage from '../index';
import history from '@app/utils/history';
import { renderWithIntl } from '@app/utils/testUtils';

describe('<NotFoundPage /> tests', () => {
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild }
    } = renderWithIntl(
      <I18nProvider i18n={i18n}>
        <NotFoundPage />
      </I18nProvider>
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should take the user back to the homePage if the go back button is clicked', () => {
    const { getByTestId } = renderWithIntl(
      <I18nProvider i18n={i18n}>
        <NotFoundPage />
      </I18nProvider>
    );
    const spy = jest.spyOn(history, 'push');
    fireEvent.click(getByTestId('back-button'));
    expect(spy).toBeCalled();
  });
});
