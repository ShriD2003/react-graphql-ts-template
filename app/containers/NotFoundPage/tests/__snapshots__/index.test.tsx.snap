// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotFoundPage /> tests should render and match the snapshot 1`] = ` `;
